File: RayTracerFSharp/RayTracer.fs
module RayTracerFSharp.Tracer

open GlmNet
open Tracer

let nonZero (vector:vec3) = vector.x <> 0.0f || vector.y <> 0.0f || vector.z <> 0.0f

let zeroVector = new vec3(0.0f, 0.0f, 0.0f)

let Trace (rayorig:vec3) (raydir:vec3) (depth:int) (sceneObjects:ResizeArray<SceneObject>) =

    let lights = sceneObjects |> Seq.filter (fun o -> o.GetMaterial(zeroVector).emissive |> nonZero)
    
    sceneObjects
    |> Seq.filter (fun sceneObject -> sceneObject.Intersects(rayorig, raydir))
    |> Seq.head
    |> (fun sceneObject -> 
        let distance = sceneObject.IntersectDistance(rayorig, raydir)
        let pointOfIntersection = rayorig + (raydir * distance)

        let material = sceneObject.GetMaterial(pointOfIntersection)
        let normal = sceneObject.GetSurfaceNormal(pointOfIntersection)

        let intensity = lights |> Seq.sumBy ( fun lightSource ->
            let lightDirection = lightSource.GetRayFrom(pointOfIntersection)
            let lightIntensity = glm.dot(normal, lightDirection)
            if lightIntensity < 0.0f then 0.0f else lightIntensity)

        material.albedo * intensity)

File: RayTracerCSharp/RayTracer.cs
using System;
using System.Linq;
using System.Collections.Generic;
using GlmNet;
using Tracer;

namespace RayTracerCSharp
{
    public static class RayTracer {

        private static vec3 ZERO_VECTOR = new vec3(0, 0, 0);

        private static bool Zero(vec3 vector) {
            return vector.x == 0 && vector.y == 0 && vector.z == 0;
        }

        // Trace a ray into the scene, return the accumulated light value
        public static vec3 TraceRay(vec3 rayorig, vec3 raydir, int depth, List<SceneObject> sceneObjects) {

            var lights = sceneObjects.Where(o => !Zero(o.GetMaterial(ZERO_VECTOR).emissive));

            foreach (var sceneObject in sceneObjects) {

                if (sceneObject.Intersects(rayorig, raydir)) {

                    float distance = sceneObject.IntersectDistance(rayorig, raydir);
                    var pointOfIntersection = rayorig + (raydir * distance);

                    var material = sceneObject.GetMaterial(pointOfIntersection);
                    var normal = sceneObject.GetSurfaceNormal(pointOfIntersection);

                    var intensity = lights.Sum(lightSource => {
                        var lightDirection = lightSource.GetRayFrom(pointOfIntersection);
                        var lightIntensity = glm.dot(normal, lightDirection);
                        return lightIntensity < 0 ? 0 : lightIntensity;
                    });

                    var pixelColour = material.albedo * intensity;
                    return pixelColour;
                }
            }

            throw new Exception("Ray didn't hit anything");
        }
    }
}
